### ISSUE TITLE
Need to serialize/deserialize the cache

### ISSUE THREAD

**Issue by Stargateur** [jaemk/cached#49]
I need to serialize the cache:

 1. There is currently no way to iterate with Cached trait
 2. Without GATs I don't think we could have an associate Iter type on Cached, maybe it's possible I don't known
 3. We could implement Serialize/Deserialize from serde for CachedSized & Co, this require some work and have some cons but this could allow to keep time(not very useful for timed cache) and order information.
 4. We could add a method in Cached to return an HashMap of Cache consuming self, not all implementation could do it without allocate a new hashmap and we lost information about time or order.

**Comment by Stargateur** [comment]
After testing around I think both could be nice to have.

**Comment by jaemk** [comment]
In order of my thoughts:

(3): I think we should avoid implementing Serialize/Deserialize since we'd have to maintain that contract between library versions and that'd be a major pain.

(2): I'm not sure either, but seems like it might be more complicated than it's worth

(1. & 4): I think we should add a `.iter` method to the Cached trait for returning all cached keys and values in an unspecified order (as if you were pulling the values from a hashmap), and then implement `std::iter::FromIterator` for each cache store. I think it's reasonable and okay to lose internal metadata (like time) when going to/from an iterator, and leave that as an implementation detail of each cache type, e.g. SizedCache returns an iterator in the order of oldest to newest and its FromIterator implementation assumes the same oldest to newest ordering so it can simply `cache_set` all the values in order.

**Comment by Stargateur** [comment]
> I think we should add a .iter method to the Cached trait for returning all cached keys and values

Yes that my first though but that not easy as its look see https://github.com/rust-lang/rust/issues/44265. I don't think it's currently possible.

**Comment by jaemk** [comment]
Hi @Stargateur,

`DiskCache` (backed by redb) and `RedisCache` both serialize/deserialize via serde, enabling
persistent caches that survive process restarts. Closing.
